gobblin-temporal/src/main/java/org/apache/gobblin/temporal/yarn/YarnServiceDynamicScalingManager.java

```diff
@@ -0,0 +1,103 @@
+  private final String DYNAMIC_SCALING_DIRECTIVES_DIR = DYNAMIC_SCALING_PREFIX + "directives.dir";
+  private final String DYNAMIC_SCALING_ERRORS_DIR = DYNAMIC_SCALING_PREFIX + "errors.dir";
+  private final String DYNAMIC_SCALING_INITIAL_DELAY = DYNAMIC_SCALING_PREFIX + "initial.delay";
+  private final int DEFAULT_DYNAMIC_SCALING_INITIAL_DELAY_SECS = 60;
+  private final String DYNAMIC_SCALING_POLLING_INTERVAL = DYNAMIC_SCALING_PREFIX + "polling.interval";
+  private final int DEFAULT_DYNAMIC_SCALING_POLLING_INTERVAL_SECS = 60;
+  private final Config config;
+  private final YarnService yarnService;
+  private final ScheduledExecutorService dynamicScalingExecutor;
+  private final FileSystem fs;
+
+  public YarnServiceDynamicScalingManager(GobblinTemporalApplicationMaster appMaster) {
+    this.config = appMaster.getConfig();
+    this.yarnService = appMaster.get_yarnService();
+    this.dynamicScalingExecutor = Executors.newSingleThreadScheduledExecutor(
+        ExecutorsUtils.newThreadFactory(Optional.of(log), Optional.of("DynamicScalingExecutor")));
+    this.fs = appMaster.getFs();
+  }
+
+  @Override
+  protected void startUp() {
+    int scheduleInterval = ConfigUtils.getInt(this.config, DYNAMIC_SCALING_POLLING_INTERVAL,
+        DEFAULT_DYNAMIC_SCALING_POLLING_INTERVAL_SECS);
+    int initialDelay = ConfigUtils.getInt(this.config, DYNAMIC_SCALING_INITIAL_DELAY,
+        DEFAULT_DYNAMIC_SCALING_INITIAL_DELAY_SECS);
+
+    Config baselineConfig = ConfigFactory.empty();
+    // TODO : Add required configs like initial containers, memory, cores, etc..
+    WorkforcePlan workforcePlan = new WorkforcePlan(baselineConfig, 0);
+    ScalingDirectiveSource scalingDirectiveSource = new FsScalingDirectiveSource(
+        this.fs,
+        this.config.getString(DYNAMIC_SCALING_DIRECTIVES_DIR),
+        java.util.Optional.of(this.config.getString(DYNAMIC_SCALING_ERRORS_DIR))
+    );
+
+    log.info("Starting the " + YarnServiceDynamicScalingManager.class.getSimpleName());
+    log.info("Scheduling the dynamic scaling task with an interval of {} seconds", scheduleInterval);
+
+    this.dynamicScalingExecutor.scheduleAtFixedRate(
+        new YarnDynamicScalingRunnable(this.yarnService, workforcePlan, scalingDirectiveSource),
+        initialDelay, scheduleInterval, TimeUnit.SECONDS
+    );
+  }
+
+  @Override
+  protected void shutDown() {
+    log.info("Stopping the " + YarnServiceDynamicScalingManager.class.getSimpleName());
+    ExecutorsUtils.shutdownExecutorService(this.dynamicScalingExecutor, Optional.of(log));
+  }
+
+  @AllArgsConstructor
+  static class YarnDynamicScalingRunnable implements Runnable {
+    private final YarnService yarnService;
+    private final WorkforcePlan workforcePlan;
+    private final ScalingDirectiveSource scalingDirectiveSource;
+
+    @Override
+    public void run() {
+      WorkforceStaffing workforceStaffing = workforcePlan.getStaffing();
```

Comment:
instead of reaching into the `WorkforcePlan`, we'll need to track our own separate/independent `WorkforceStaffing` to reflect currently requested/allocated containers -- "actual" vs. what the directives indicate to be "planned".  (DO keep in mind there may be nuance to handling the difference between in-progress requested and already allocated.)

avoid doing too much here in this class: keep it to polling the directory for `ScalingDirectives`.  (perhaps maintain a timestamp of "latest/newest" directive so you'll know when something new has arrived vs when you've processed all the directives already.)

meanwhile the plan revision and rescaling can live in a `YarnService` extension(*).  it's the one requesting the containers and also getting updates about when containers fail or are preempted, and up-to-date staffing will be easier to maintain when co-located w/ staffing level mgmt/updates.  hence `YarnService` would contain both the `WorkforcePlan` and the "current/actual" `WorkforceStaffing` levels.  this DSMgr would merely pass scaling directives to that "`YarnService`".

(*) rather than adding a bunch of dynamic scaling concerns to bloat the `YarnService` impl, let's extend that class (`DynamicScalingYarnService` maybe?), to separately address and centralize these new concerns there.